File: cases/024.metric-recorder-buffered-flush/src/metric_collector.cc
#include "metric_collector.h"

namespace nitr::case024 {

MetricCollector::MetricCollector(MetricRecorder& recorder) : recorder_(recorder) {}

void MetricCollector::Collect(const std::vector<double>& samples) {
  if (samples.empty()) {
    recorder_.Record({"samples.count", 0.0});
    return;
  }

  double sum = 0.0;
  double max_value = samples.front();
  for (double v : samples) {
    sum += v;
    if (v > max_value) max_value = v;
  }

  const double count = static_cast<double>(samples.size());
  recorder_.Record({"samples.count", count});
  recorder_.Record({"samples.mean", sum / count});
  recorder_.Record({"samples.max", max_value});
}

}  // namespace nitr::case024

File: cases/032.metric-recorder-buffered-flush-python/src/metric_collector.py
from __future__ import annotations

from src.metric_recorder import Metric, MetricRecorder


class MetricCollector:
    def __init__(self, recorder: MetricRecorder) -> None:
        self._recorder = recorder

    def collect(self, samples: list[float]) -> None:
        if not samples:
            self._recorder.record(Metric("samples.count", 0.0))
            return

        sample_count = float(len(samples))
        sample_sum = sum(samples)
        max_value = max(samples)

        self._recorder.record(Metric("samples.count", sample_count))
        self._recorder.record(Metric("samples.mean", sample_sum / sample_count))
        self._recorder.record(Metric("samples.max", max_value))
